### ISSUE TITLE
Questions regarding implementation

### ISSUE THREAD

**Issue by anthony0727** [lucidrains/dreamerv3-pytorch#3]
I've been implementing Dreamer V3 in rllib and can't get it working quite yet,
and I have few questions

in the paper, at **Symlog Prediction**

> DreamerV3 uses symlog predictions in the decoder, the reward predictor, and the critic. It also squashes inputs to encoder~

at **World Model Learning**

> The prediction loss trains the decoder and reward predictor via the symlog loss~

Does symlog loss mean that,
we calculate respective losses using `symlog(the outputs)`?

I understand the critic part, but

squashing inputs to encoder,
`latent = self.model.encoder(symlog(train_batch["obs"]))`

and using the `latent` to calculate the followings,
and then losses using `symlog(the outputs)` again seems too many symlog transformation

Thanks ahead!
Anthony

**Comment by danijar** [comment]
Hey, for the decoder predictions, symlog is applied to the targets (i.e. the data) not the predictions.

**Comment by anthony0727** [comment]
yup, then is targets only part symlog is applied to??

**Comment by anthony0727** [comment]
lambda returns consist of data and predictions i.e. reward + value_prediction

but especially critic loss part from the paper, it seems symlog is applied to lambda return

can you exactly describe how many symlog is applied to where?

from my understanding, it's applied to
1. raw observation 2. reward from env 3. lambda return

Thanks for the reply

**Comment by danijar** [comment]
Sure, happy to clarify.

- Encoder: Simply preprocess non-image inputs to the encoder using symlog. It's not needed for images because they are already in range [0, 1] where symlog has little effect.
- Decoder: The decoder outputs are trained to predict not the raw observations/rewards/lambda-targets `x` but the symlog compressed observations `symlog(x)`.
- Reward predictor: The reward predictor outputs are trained to predict not the raw rewards `r` but the symlog compressed rewards `symlog(r)`.
- Critic network: The critic outputs are trained to predict not the raw lambda targets `R_lambda` but the symlog compressed lambda targets `symlog(R_lambda)`.

Importantly, when using the predictions of the decoder, reward predictor, or critic network, the neural network outputs `f(h, z)` are mapped back into the uncompressed space to read out their predictions `y_pred = symexp(f(h, z))`.

Additionally, the reward predictor and critic network are trained using a two-hot categorical regression loss instead of the more common MSE. For this, the network outputs `f(h, z)` a vector of 255 logits that is trained to predict the two-hot encoding of the targets. Again, we need to "undo" this idea when reading out the network predictions. In this case, this works by taking the inner product between the predicted bin probabilities and the bins positions `softmax(f(h, z)).T @ bin_values`. This doesn't apply to the decoder, where we use MSE because it's potentially predicting many scalar quantities and outputting logits for all those dimensions might end up using too much memory/compute.

Symlog predictions and two-hot categorical regression are orthogonal techniques. But because we're using both, the transformations are chained when computing training targets `y' = twohot(symlog(y))` and the inverse steps are also chained when reading out predictions of the network `y_pred = symexp(softmax(f(h, z)).T @ bin_values)`. And the loss function is `L = categorical_cross_entropy(logits=f(h, z), targets=y')`.

**Comment by anthony0727** [comment]
Thanks a million for thorough explanation!

It seems that I missed few symexp inverse step

Lastly, for the percentiled-rescaled lambda return of actor loss, the lambda return is algo symexp-ed right?

**Comment by danijar** [comment]
The lambda return is computed from symexp'ed rewards and critic predictions. The lambda return is then symlog'ed for computing the critic loss. But the raw lamba return is used for computing the actor loss, otherwise it would introduce a lot of bias. The lambda return is never symexp'ed.

**Comment by sdpmas** [comment]
@danijar any plans to release the code soon?

**Comment by anthony0727** [comment]
is the bin boundary [-20, ..., 20] as depicted in the paper?
